### ISSUE TITLE
Expose setting xres/yres for JPEG/PNG output

### ISSUE THREAD

**Issue by Andriy-Kulak** [lovell/sharp#967]
Hi!

I saw that in the documentation it is possible to adjust the density metadata in the constructor, but it doesn't work in my case for some reason. Below is the code which works but the density is never adjusted. Does anyone know how I can do this with the code snippet below?

Thanks!

```
const sharpImage = sharp(imagedData /* tried adding density here ... , {density: 170} */)
      sharpImage
      .rotate()
      .metadata()
      .then(function(metadata) {
        return sharpImage
                .extract({
                  left: Math.round(metadata.width * parasedBody.x),
                  top: Math.round(metadata.height * parasedBody.y),
                  width: Math.round(metadata.width * parasedBody.width),
                  height: Math.round(metadata.height * parasedBody.height)})
                .toBuffer((err, outputBuffer, info) => {
                  console.log("BUFF FUNC ERR", err);
                  console.log("BUFF FUNC outputBuffer", outputBuffer);
                  console.log("BUFF FUNC info", info);
                })
        })
```

**Comment by lovell** [comment]
Hello, `density` only applies to vector-based input such as SVG. What input format are you using?

**Comment by Andriy-Kulak** [comment]
I allow user to only upload a PNG/JPEG type of image. I then convert it to base64 image (which is labeled "imageData" above) and then do the cropping as seen above. I need to set a specific density/DPI/resolution in the metadata so that when I interact with a third party API that does prints, their API knows exactly how many pixels per inch to print. By default the sharp module changes the density to 72 (as explained in documentation and seen in image attached). I assume that if is changing the density from whatever it was before to 72, then I can customize this metadata.

Is this possible to change this metadata?

Thanks!
<img width="400" alt="screen shot 2017-09-29 at 1 42 33 pm" src="https://user-images.githubusercontent.com/12179976/31028573-75929a3a-a51c-11e7-9c11-cc2fd391bc5c.png">

**Comment by lovell** [comment]
Did you see https://github.com/lovell/sharp/issues/823#issuecomment-332742955 ?

**Comment by Andriy-Kulak** [comment]
@lovell Yes, so when the image is Jpeg, I cannot add a resolution/DPI to it. However, once I convert it to a base 64 image, is it still not possible to add a resolution metatag? If it is automatically set to 72 once I am done cropping, maybe I can change it to 170 which is what I need. I don't have in depth knowledge of image processing but if you can give a hint on where that is being done within sharp code, I can take a look.

**Comment by Andriy-Kulak** [comment]
This is the metadata coming from the JPG image in line 5 of the code i provided above. I understand that density/resolution doesn't mean anything for JPG in the digital sense per se, but is it possible to change the metadata here so that when the jpeg is used for printing, it will be used by the third party API I have to work with?

Thanks!

```
{ format: 'jpeg',
width: 3840,
height: 2560,
space: 'srgb',
channels: 3,
density: 72,
hasProfile: false,
hasAlpha: false }
```

**Comment by lovell** [comment]
The presence of `density` in the output of `metadata` for a JPEG input image is probably a bug. It's not derived from actual metadata, sorry for the confusion.

**Comment by Andriy-Kulak** [comment]
So is there any possibility to add density in some sort of way via Sharp to JPEG/PNGs in the future? I found that in imagemagick it is possible per link below, but I don't want to run a whole instance of ImageMagick just to add one piece of metadata. If you point me in the right direction, I can try to help implement this.

https://superuser.com/questions/479197/i-want-to-change-dpi-with-imagemagick-without-changing-the-actual-byte-size-of-t

**Comment by lovell** [comment]
The addition of custom EXIF metadata is covered by #650.

**Comment by lovell** [comment]
Ah, I see what's happening here, the input image probably has a JFIF metadata block. Sorry for the confusion (I'm confusing myself).

After the work for #650 to make the existing `withMetadata` more powerful, exposing the ability to set libvips' internal `xres` and `yres` fields on the `jpeg()` function (like is already available on the `tiff()` function) is probably the best way to achieve what is needed here.

**Comment by lovell** [comment]
See https://github.com/lovell/sharp/issues/823#issuecomment-333464057

**Comment by lovell** [comment]
Cross-post at https://stackoverflow.com/questions/46508181/how-to-set-a-resolution-density-in-jpeg-png-image-in-javascript

**Comment by Andriy-Kulak** [comment]
Yes, thats my post lol. I was able to get it to work in node's version of image magic. I had trouble getting it to work with [piexifjs](https://github.com/hMatoba/piexifjs) package for some reason. I will be away the next two weeks, but afterwards I will look into implementation in sharp.

Thanks!

**Comment by borie88** [comment]
Any updates to this? I'm also using this for printing purposes, thanks

**Comment by jasonmacdonald** [comment]
Funny, I'm having the opposite problem. A JPG uploaded with the DPI of 300, is still 300 when I resize it and save it. :( I need it to convert to 72 but it doesn't seem to work.

**Comment by giorgionetg** [comment]
Currently, as in Gimp/Photoshop, the density it's only a parameter, as parameter does not change size resolution!* With 300 DPI or 72 DPI of 300*200 image we have always 300*200 pixel at all. I think the library here, get the allready exist parameter inside image.

There are two ways to solve this:

- Working with fixed DPI (like as 72 DPI default) and change resolution by itself (with size and resize). More easy to implement.
- Working as Gimp/Photoshop, but this means a lot of work on library.

* I know in Gimp/Photoshop if you change DPI you can change resolutions, but nothing to much than you can do by yourself as first way I shown.

**Comment by lovell** [comment]
Commit https://github.com/lovell/sharp/commit/4237f5520f6dfcae2706e6d7ffe3cc3d3e05a604 adds a new output `density` property to `withMetadata()` to set the DPI resolution, which keeps things consistent with the input `density` property of `metadata()`.

```js
// Available from v0.28.2

const data = await sharp(input)
  .withMetadata({ density: 96 })
  .toBuffer();
```

This will be in v0.28.2.

**Comment by lovell** [comment]
v0.28.2 is now available.
